### ISSUE TITLE
Broken instructions in docs/howto-setup-exercise-development-environment.md?

### ISSUE THREAD

**Issue by fpottier** [ocaml-sf/learn-ocaml#203]
Step 3 in the instructions seems broken: the command `learn-ocaml` is available inside `$DIR/learn-ocaml` (because we have created a local switch), but not inside `$DIR`.

I have tried to work around the problem by working directly inside `$DIR/learn-ocaml`,
as follows:
```
master $ learn-ocaml build --repo demo-repository
Learnocaml v.0.9 running.
Updating app at ./www
demo                         [OK]
master $ learn-ocaml serve
Learnocaml server v.0.9 starting on port 8080
Initial teacher token created: X-OLL-ACF-9G1-JTO
2018-12-04T20:36:18Z	127.0.0.1	GET 	/
2018-12-04T20:36:18Z	127.0.0.1	GET 	/favicon.ico
```

But then, when I try to access `http://localhost:8080`, I get a page that contains the following error:
```
Error: Unix.Unix_error(Unix.ENOENT, "open", "./www/index.html")
```

**Comment by Armael** [comment]
>the command learn-ocaml is available inside $DIR/learn-ocaml (because we have created a local switch), but not inside $DIR.

My understanding is that once you do `eval $(opam env)` while in the learn-ocaml directory, this will setup your shell environment using the local switch and give you access to the learn-ocaml binary. Then, `learn-ocaml` should still be accessible even if you go in `$DIR`, as long as you stay in the same shell.

**Comment by fpottier** [comment]
Did not work for me. But I have noticed that `eval $(opam env)` is broken (it has no effect) since I upgraded to opam 2, so perhaps that is indeed the source of my problem.

**Comment by rjbou** [comment]
This is weird...
Did you installed shell hooks? Is `opam env` result empty?

**Comment by fpottier** [comment]
Here is a sample interaction, where I switch from 4.07.0 to 4.06.1,
and `eval $(opam env)` does not have the desired effect:

```
$ opam --version
2.0.0
$ ocamlc -v
The OCaml compiler, version 4.07.0
Standard library directory: /home/fpottier/.opam/4.07.0/lib/ocaml
$ opam switch 4.06.1
$ opam env
OPAM_SWITCH_PREFIX='/home/fpottier/.opam/4.06.1'; export OPAM_SWITCH_PREFIX;
$ eval $(opam env)
$ ocamlc -v
The OCaml compiler, version 4.07.0
Standard library directory: /home/fpottier/.opam/4.07.0/lib/ocaml
```

Strangely enough, if I switch back to 4.07.0, `opam env` prints out a lot more settings:

```
$ opam switch 4.07.0
$ opam env
OPAM_SWITCH_PREFIX='/home/fpottier/.opam/4.07.0'; export OPAM_SWITCH_PREFIX;
CAML_LD_LIBRARY_PATH='/home/fpottier/.opam/4.07.0/lib/stublibs:/home/fpottier/.opam/4.07.0/lib/ocaml/stublibs:/home/fpottier/.opam/4.07.0/lib/ocaml'; export CAML_LD_LIBRARY_PATH;
OCAML_TOPLEVEL_PATH='/home/fpottier/.opam/4.07.0/lib/toplevel'; export OCAML_TOPLEVEL_PATH;
MANPATH=':/home/fpottier/.opam/4.07.0/man:/home/fpottier/.opam/4.07.0/man'; export MANPATH;
PATH='/home/fpottier/.opam/4.07.0/bin:/home/fpottier/bin:/home/fpottier/.opam/4.07.0/bin:/home/fpottier/bin:/home/fpottier/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin'; export PATH;
```

I have observed this kind of bizarre behavior both under Ubuntu and under MacOS.

**Comment by rjbou** [comment]
Indeed, there is something broken with the opam env... Maybe better move further discussions on [opam bug tracker](https://github.com/ocaml/opam/issues). Can you open an issue?

Also:
-  does theses switches are opam 1.2 imported or opam 2.0 created?
- an you share the files `/home/fpottier/.opam/4.0{6,7}.0/.opam-switch/environment`

**Comment by fpottier** [comment]
Thanks, Raja, I will open an issue.
I just upgraded from opam 2.0.0 to opam 2.0.1 and now still have problems with `opam env`, although different problems.

Regarding the current issue, I still think that the instructions in `docs/howto-setup-exercise-development-environment.md` possibly do not work. E.g., the user may have enabled opam's environment hook, which means that `eval $(opam env)` is automatically executed after every command.

**Comment by yurug** [comment]
@AltGr, can you help us? Is opam's environment hook incompatible with a switch specific to learn-ocaml?

**Comment by AltGr** [comment]
@yurug absolutely not: on the contrary, it should help a lot, and that's what the hook is there for.

- @fpottier's issue with `opam env` will be discussed on opam's tracker, but I think it's due to a quirk in the upgrade from previous opam versions. In any case, recreating the switch should solve it; it may be due to a broken `ocaml` package in the switch, please check `_opam/.opam-switch/config/ocaml.config` (which should contain e.g. the `native-dynlik` variable). You could also try to rebuild `_opam/.opam-switch/environment` by removing it and reinstalling any package.

- in any case, that issue affects both `opam env` and the shell hook in the same way, and is not related to learn-ocaml.

- the `eval $(opam env)` is just unneeded if the shell hooks have been set up. It won't change anything, so it's probably better to leave it there to handle all cases, and avoid a complex explanation of the cases where you might or not need it.

**Comment by yurug** [comment]
@AltGr, thanks for the clarification!

**Comment by yurug** [comment]
@fpottier, about your failed attempt to run `learn-ocaml`: did you run the command from the root of the demo-repository? Otherwise, it is probably normal that `learn-ocaml` cannot find `./www`.

**Comment by fpottier** [comment]
Well, the `howto` suggests running `learn-ocaml` *above* the demo repository, not *inside* it. Anyway, I tried both, without any luck.

**Comment by fpottier** [comment]
Futhermore, I think I disagree with @AltGr's answer. If the user executes an implicit `eval $(opam env)` after each command, then when he executes `cd $DIR`, he loses the command `learn-ocaml`, which is no longer in the PATH. So the instructions do not work.

**Comment by EmileRolley** [comment]
Closed by #221
